src/net.h

```diff
@@ -1257,6 +1276,40 @@ class CConnman
      */
     std::vector<CService> m_onion_binds;
 
+    /**
+     * RAII helper to atomically create a copy of `vNodes` and add a reference
+     * to each of the nodes. The nodes are released when this object is destroyed.
+     */
+    class NodesSnapshot
+    {
+    public:
+        explicit NodesSnapshot(const CConnman& connman) : m_connman{connman}
+        {
+            LOCK(m_connman.cs_vNodes);
+            m_nodes_copy = m_connman.vNodes;
+            for (auto& node : m_nodes_copy) {
+                node->AddRef();
+            }
+        }
+
+        ~NodesSnapshot()
+        {
+            LOCK(m_connman.cs_vNodes);
+            for (auto& node : m_nodes_copy) {
+                node->Release();
```

Comment:
I see that you haven't changed the behavior here, but I'm curious about the pre-existing logic: 

Is it necessary for `Release()` to be under the lock? 

It makes sense that `AddRef` has to be under the lock so there isn't a race between a thread that is trying to operate on the node, and a thread that is evaluating for disconnect. However, with `Release`, I don't think there is actually potential for a race.. if `GetRefCount` returns > 0, `DisconnectNodes` will delay processing it. And since `nRefCount` is atomic, it shouldn't be a problem if multiple threads are trying to increment / decrement at the same time. Does this reasoning seem sound? Am I missing something?

Reply:
I think this is correct. There's no need to hold `cs_vNodes` when calling `node->Release();`

Removing this lock means that the `m_connman` member is no longer required.

Reply:
> Is it necessary for `Release()` to be under the lock?

The lock has two effects:
* All `Release()`s happen atomically together. If removed, then other threads could observe some nodes from the snapshot released and some not yet released. I think that is ok.
* No memory reordering can happen. The `nRefCount--` on the atomic variable has the same effect because `--` is equivalent to [`fetch_sub()`](https://en.cppreference.com/w/cpp/atomic/atomic/fetch_sub) which uses [`memory_order_seq_cst`](https://en.cppreference.com/w/cpp/atomic/memory_order) by default.


Removed the lock and simplified `NodesSnapshot`, thanks!